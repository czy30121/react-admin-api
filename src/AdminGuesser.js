import React from 'react';
import PropTypes from 'prop-types';
import {Admin, Loading, TranslationProvider, Query} from 'react-admin';
import {createMuiTheme} from '@material-ui/core/styles';
import ResourceGuesser from './ResourceGuesser';
import {notExistsAsChild} from './helpers';

const theme = createMuiTheme({
  palette: {
    primary: {
      main: '#38a9b4',
      contrastText: '#fff',
    },
    secondary: {
      main: '#288690',
    },
  },
});

const notBlackListed = (blacklist, resourceName) => {
  const blacklisted = new Set(blacklist);

  return !blacklisted.has(resourceName);
};

const AdminGuesser = ({children, blacklist, ...props}) => (
  <Query type="INTROSPECT">
    {({data, loading, error}) => {
      if (loading) {
        return (
          <TranslationProvider>
            <Loading />
          </TranslationProvider>
        );
      }
      if (error) {
        console.error(error);
        return <div>Error while reading the API schema</div>;
      }

      return (
        <Admin theme={theme} {...props}>
          {children || <></>}
          {data.resources
            .filter(
              resource =>
                !resource.deprecated &&
                notBlackListed(blacklist, resource.name) &&
                notExistsAsChild(children, 'name', resource.name),
            )
            .map(resource => (
              <ResourceGuesser name={resource.name} key={resource.name} />
            ))}
        </Admin>
      );
    }}
  </Query>
);

AdminGuesser.propTypes = {
  blacklist: PropTypes.array,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
};

export default AdminGuesser;
