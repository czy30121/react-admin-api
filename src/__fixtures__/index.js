import resources from "./resources";

export {
  resources
};
