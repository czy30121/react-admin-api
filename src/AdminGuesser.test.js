import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import {createStore} from 'redux';
import {Provider} from 'react-redux';
import {render} from './AdminGuesser';
import {resources} from "./__fixtures__";

describe('<AdminGuesser />', () => {
  const renderer = new ShallowRenderer();
  const store = createStore((state = {}) => state);

  test('renders errors', () => {
    const tree = renderer.render(
      <Provider store={store}>
        {render({}, {error: 'Failed to fetch documentation', loading: false})}
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  test('renders loading', () => {
    const tree = renderer.render(
      <Provider store={store}>{render({}, {loading: true})}</Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  test('renders without custom resources', () => {
    const tree = renderer.render(
      <Provider store={store}>
        {render({}, {data: {resources}, loading: false})}
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });
});
