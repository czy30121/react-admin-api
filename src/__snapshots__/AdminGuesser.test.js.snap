// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminGuesser /> renders loading 1`] = `
<Loading
  loadingPrimary="ra.page.loading"
  loadingSecondary="ra.message.loading"
/>
`;

exports[`<AdminGuesser /> renders with custom resources 1`] = `
<AdminUI
  catchAll={[Function]}
  layout={[Function]}
  loading={[Function]}
  loginPage={[Function]}
  logout={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "propTypes": Object {
        "className": [Function],
        "icon": [Function],
        "redirectTo": [Function],
      },
      "render": [Function],
    }
  }
>
  <ResourceGuesser
    name="custom"
  />
</AdminUI>
`;

exports[`<AdminGuesser /> renders without custom resources 1`] = `
<AdminUI
  catchAll={[Function]}
  layout={[Function]}
  loading={[Function]}
  loginPage={[Function]}
  logout={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "propTypes": Object {
        "className": [Function],
        "icon": [Function],
        "redirectTo": [Function],
      },
      "render": [Function],
    }
  }
>
  <ResourceGuesser
    name="books"
  />
</AdminUI>
`;

exports[`<AdminGuesser /> renders without data 1`] = `
<AdminUI
  catchAll={[Function]}
  layout={[Function]}
  loading={[Function]}
  loginPage={[Function]}
  logout={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "propTypes": Object {
        "className": [Function],
        "icon": [Function],
        "redirectTo": [Function],
      },
      "render": [Function],
    }
  }
/>
`;
